lib/net/ssh/transport/packet_stream.rb

```diff
@@ -243,26 +263,38 @@ def poll_next_packet
           if need > 0
             # read the remainder of the packet and decrypt it.
             data = read_available(need)
-            @mac_data += data if server.hmac.etm
-            @packet.append(server.update_cipher(data))
+            if server.hmac.aead
+              @packet.append(data)
+            else
+              @mac_data += data if server.hmac.etm
+              @packet.append(server.update_cipher(data))
+            end
           end
 
           # get the hmac from the tail of the packet (if one exists), and
           # then validate it.
           real_hmac = read_available(server.hmac.mac_length) || ""
 
-          @packet.append(server.final_cipher)
-          padding_length = @packet.read_byte
-
-          payload = @packet.read(@packet_length - padding_length - 1)
+          if server.hmac.aead
```

Comment:
FWIW This is what the chachapoly version looks like:

https://github.com/net-ssh/net-ssh/blob/5bd322359df9d6a7cec8e0ee1e95437564adcd33/lib/net/ssh/transport/packet_stream.rb#L265-L269